Main.hs

```diff
@@ -18,29 +18,99 @@ import qualified Network.Wai as Wai
 import qualified Network.Wai.Handler.Warp as Warp
 import           Network.Wai.Handler.WebSockets (websocketsOr)
 import           Network.WebSockets.Connection (defaultConnectionOptions)
+import           Routes
 import qualified Sockets
 import           System.Random (getStdRandom, randomR)
 import           Web.Scotty
 
-
 maybeParam :: Parsable a => Lazy.Text -> ActionM (Maybe a)
 maybeParam name = (Just <$> param name) `rescue` (return . const Nothing)
 
-getHeader :: HeaderName -> ActionM (Maybe Strict.Text)
-getHeader name = (listToMaybe . map (Strict.decodeUtf8 . snd)
-                  . filter ((== name) . fst) . Wai.requestHeaders) <$> request
+maybeHeader :: HeaderName -> ActionM (Maybe Strict.Text)
+maybeHeader name = (listToMaybe . map (Strict.decodeUtf8 . snd)
+                   . filter ((== name) . fst) . Wai.requestHeaders) <$> request
+
+getHeader :: HeaderName -> ActionM Strict.Text
+getHeader name = maybe (raise message) return =<< maybeHeader name
+  where message = "missing \"" <> headerName <> "\" header"
+        headerName = (Lazy.fromStrict . Strict.decodeUtf8 . original) name
+
+route :: Database
+      -> Chan (EmailAddress, Code)
+      -> Chan ResolvedPost
+      -> (ActionM () -> ScottyM ())
+      -> ActionM (Either Response Request)
+      -> ScottyM ()
+route db emailChan socketChan path makeReq = path $ do
+  reqOrRes <- makeReq
+  let dbRes = either return execute reqOrRes
+  res <- liftIO (runReaderT dbRes db)
+  effects <- send res
+  mapM_ (liftIO . perform) effects
```

Comment:
@hlian Didn't you say `mapM` is bad news? What's the correct alternative? `traverse`?